toolchain/sem_ir/inst_namer.cpp

```diff
@@ -355,27 +356,45 @@ auto InstNamer::AddBlockLabel(ScopeId scope_id, SemIR::LocId loc_id,
   AddBlockLabel(scope_id, branch.target_id, name.str(), loc_id);
 }
 
-// TODO: Consider addressing recursion here. It may be important because
-// InstNamer is used for debug info.
-// NOLINTNEXTLINE(misc-no-recursion)
 auto InstNamer::CollectNamesInBlock(ScopeId scope_id, InstBlockId block_id)
     -> void {
   if (block_id.is_valid()) {
     CollectNamesInBlock(scope_id, sem_ir_->inst_blocks().Get(block_id));
   }
 }
 
-// NOLINTNEXTLINE(misc-no-recursion): See above TODO.
-auto InstNamer::CollectNamesInBlock(ScopeId scope_id,
+auto InstNamer::CollectNamesInBlock(ScopeId top_scope_id,
                                     llvm::ArrayRef<InstId> block) -> void {
-  Scope& scope = GetScopeInfo(scope_id);
+  std::deque<std::pair<ScopeId, InstId>> insts;
+
+  // Appends a scope and instructions to walk. Avoids recursion while allowing
+  // the loop to below append more instructions during iteration.
+  auto append_block_id = [&](ScopeId scope_id, InstBlockId block_id) {
+    if (block_id.is_valid()) {
+      for (auto inst_id : sem_ir_->inst_blocks().Get(block_id)) {
+        insts.emplace_back(scope_id, inst_id);
+      }
+    }
+  };
+  auto append_block_insts = [&](ScopeId scope_id,
+                                llvm::ArrayRef<InstId> inst_ids) {
+    for (auto inst_id : inst_ids) {
+      insts.emplace_back(scope_id, inst_id);
+    }
+  };
```

Comment:
Regarding the use of deque, had you considered instead using SmallVector? I think that just requires adding in reverse order, such as:

```suggestion
  llvm::SmallVector<std::pair<ScopeId, InstId>> insts;

  // Appends a scope and instructions to walk. Avoids recursion while allowing
  // the loop to below append more instructions during iteration.
  auto append_block_id = [&](ScopeId scope_id, InstBlockId block_id) {
    if (block_id.is_valid()) {
      for (auto inst_id : llvm::reverse(sem_ir_->inst_blocks().Get(block_id))) {
        insts.emplace_back(scope_id, inst_id);
      }
    }
  };
  auto append_block_insts = [&](ScopeId scope_id,
                                llvm::ArrayRef<InstId> inst_ids) {
    for (auto inst_id : llvm::reverse(inst_ids)) {
      insts.emplace_back(scope_id, inst_id);
    }
  };
```

Reply:
Note: my understanding is that it'd probably be better to switch to `push_back` instead of `emplace_back` too (https://github.com/carbon-language/carbon-lang/pull/4651#pullrequestreview-2503550581)

Reply:
Switching from `deque` to a vector will switch us from FIFO to LIFO traversal -- that is, from breadth-first to depth-first -- which should keep the instruction numbering more predictable and closer to how it currently is. (We'll still have some differences because we're switching from in-order to pre-order traversal so we'll number later instructions in the parent block before we number the child block.)

Reply:
Just to note, this'd switched from emplace_back/pop_front to emplace_front/pop_front (using llvm::reverse similar to what I'd suggested here). Although that means push_back/pop_back won't change traversal, it also means the deque's FIFO feature isn't used.

Reply:
> Note: my understanding is that it'd probably be better to switch to `push_back` instead of `emplace_back` too ([#4651 (review)](https://github.com/carbon-language/carbon-lang/pull/4651#pullrequestreview-2503550581), [#4651 (comment)](https://github.com/carbon-language/carbon-lang/pull/4651#discussion_r1887759323))

Ah, I didn't think that conclusion was so general, and pair is the poster child of emplace, but I would personally be  happy to never write emplace again. :)

Originally I was pushing to the back, and popping from the front, but I changed that, so yeah a vector makes a lot more sense now, thanks for noticing that.